### DISCUSSION TITLE
For embedded pocketbase, is it possible to access the user records without spinning up the webserver?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#913] [Q&A] [answered]
I'm currently reviewing pocketbase to see if it's worth changing out most of my project skeleton to get some many free goodies.  One thing that my current system has is the ability to add and list user records from the command line, and to install the system (with embedded SQL, etc).

Can someone explain to me what happens when other Cobra commands are added and then the pocketbase app is started with `app.Start()`?  I mean, if it's an sqlite file, then can other processes access this data without it going through the long running server?  I'll start tracing the code soon to figure that out, but perhaps someone could explain to me the basic philosophy of how pocket base is embedded into an app.  I couldn't find examples of a variety of this.  In particular, when running Cobra subcommands, when does the webserver get started, and when does it not?

**Comment by ganigeorgiev** [comment, marked as answer]
You can access the database by external processes while the PocketBase server is running. The only limitation is that you can have only a single writer at a time. Writes are usually really fast and concurrent writes will just queue up (in the worst case scenario if a write transaction takes more than 10s to complete it will return `SQLITE_BUSY` for all other writes).

Another gotcha that you may face if you update the database from some other process is that **your changes will not trigger the [PocketBase event hooks](https://pocketbase.io/docs/event-hooks/) (including the realtime events) in your other PocketBase process(es)** because they are triggered only by and in the running app process.
For example, if you have a Cobra command that inserts a new record while the web server is running, you'll not receive the `OnModelBeforeCreate` and `OnModelAfterCreate` events in the web server process that a new record was added but you'll be able to query it just fine when making a new request to fetch it.

> In particular, when running Cobra subcommands, when does the webserver get started, and when does it not?

The web server starts only when you run the `serve` command (aka. `pocketbase serve`). You can see the default registered commands in https://github.com/pocketbase/pocketbase/blob/master/pocketbase.go#L125-L133.

---

_btw. If you are planning to start a new project, it may be a good idea to use the latest pre-release [v0.8.0-rc2](https://github.com/pocketbase/pocketbase/releases/tag/v0.8.0-rc2), because there are some breaking changes and it will be easier to upgrade later. More details you could find in the [rc1 changelog](https://github.com/pocketbase/pocketbase/releases/tag/v0.8.0-rc1)_

**Reply by benallfree** [reply]
> You can access the database by external processes while the PocketBase server is running

I did not know SQLite allowed concurrent access by multiple processes. I thought there was a process-based lock file. Does this mean I could potentially connect to the SQLite database from nodejs? If so, this opens up a lot of possibilities for pockethost 👯

**Reply by ganigeorgiev** [reply]
@benallfree Yes, multiple applications can access the same db file, but only one can make changes at a time. You could read more in the official SQLite FAQ - https://www.sqlite.org/faq.html#q5.

**Reply by benallfree** [reply]
That's interesting, thanks. One of my main inspirations with pockethost was to open things to web devs so they wouldn't need to learn Go to leverage all the PB goodness you've created. This could open up possibilities for writing and running migrations all from js. Perhaps even writing to collections, although that would bypass all the PB event hooks including realtime stuff, so I'd probably be better off proxying collection mutations through the PB API.

**Comment by dschissler** [comment]
@ganigeorgiev At this time I'm only interested in embedding Pocketbase.  I need to figure out how much of my app should be run by the Pocketbase webserver and how much should exist outside of it.  It seems that anything that Pocketbase controls can't be properly run through the command line unless it's accessing a Pocketbase endpoint.  So what about maintenance routines, or installation of resources?  I like that I can currently run `myprog install ...` with varying options of a partial or complete install.  This allows me to zap entire sqlite sections of the program for development and then to try again.

Would everything need to be controlled by Pocketbase?  One way around this would just be to have embedded Pocketbase only control user related stuff and everything else is handled differently as I'm doing.  But then I'd miss out on real time features.

I'm not familiar with this form of framework, so I'm just guessing about how it could be done.  Is it idiomatic to use a API login token and then use that to run a "backend" command through a Pocketbase Echo route?  This seems weird from what I'm used to, but perhaps it's fine due to Golang being so good with goroutines.

**Reply by ganigeorgiev** [reply]
> It seems that anything that Pocketbase controls can't be properly run through the command line unless it's accessing a Pocketbase endpoint. So what about maintenance routines, or installation of resources? I like that I can currently run myprog install ... with varying options of a partial or complete install. This allows me to zap entire sqlite sections of the program for development and then to try again.

I'm not sure that I understand your example. It's not required to create an API endpoint if you just want to access and insert some data into the database. You can register your own console command and access the database using the `app.Dao()` instance (more documentation will be available with the v0.8.0 release, but if you haven't already, you could check https://pocketbase.io/docs/db-queries/). The only issue is that all events will be fired and handled only within the running process, aka. there are isolated and there is no direct communication between the different running PocketBase commands.

PocketBase doesn't do anything too much unorthodox and you can use it almost as it is with any other Go toolkit/framework. Could you elaborate a little more on your use case?

**Reply by dschissler** [reply]
> The only issue is that all events will be fired and handled only within the running process, aka. there are isolated and there is no direct communication between the different running PocketBase commands.

Ok, I think that I understand now.  So if I ran another process and accessed models through app.Dao()` then I'd still have the event hooks being run (so like processing data related to the record, etc) but that it doesn't (real time) communicate the changes to the main system?  If that is the only issue, then that seems quite ok.

You've been a big help.  I'm going to dig into the code more before I ask any more questions.

**Reply by ganigeorgiev** [reply]
>  So if I ran another process and accessed models through app.Dao()` then I'd still have the event hooks being run (so like processing data related to the record, etc) but that it doesn't (real time) communicate the changes to the main system?

Yes. Technically, there is no "main" system and every command (including the `serve`) has access to more-or-less the same app internals and works independently. For example, if you have 2 commands A and B, you'll get events in A only when something is executed in A, and similarly - you'll get events in B only when something is executed in B.

**Comment by dschissler** [comment]
@ganigeorgiev Also, I'm not even a day into learning about this project, but I can't help but think that Cobra doesn't need to be an inseparable part of this project.  I mean, all of this setup stuff could be put into a function where the parameters are just passed in.  Then there could be a module for cobra integration.  This is kind of a big impediment for me because I'm already using `urfave/cli`, and actually I moved to it early from `cobra` because I like it better.  So for me it's both work, and it feels worse.

https://github.com/pocketbase/pocketbase/blob/099230a55218a3abaef3348e13abefd2af873f02/cmd/serve.go#L32-L112
